File: python_implementations/parser.py
import re


def list_rindex(li, x):
    for i in reversed(range(len(li))):
        if li[i] == x:
            return i
    raise ValueError("{} is not in list".format(x))


def parse_range(astr):
    result = set()
    for part in astr.split(','):
        x = part.split('-')
        result.update(range(int(x[0]), int(x[-1]) + 1))
    return sorted(result)


def expand(parts, stitch_symbols, expanded):
    multiplier = 1
    if expanded is None:
        expanded = []
    i = 0
    while i < len(parts):
        p = parts[i]
        internal_expanded = []
        multiplier_set = False
        if p.isdigit():
            multiplier = int(p)
            multiplier_set = True
        elif p in stitch_symbols:
            internal_expanded.append(p)
        elif p == '(':
            index = list_rindex(parts, ')')
            expand(parts[i + 1:index], stitch_symbols, internal_expanded)
            i = index
        elif p.startswith('#'):
            internal_expanded.append(f"change_color {p.strip('# ')}")

        internal_expanded = internal_expanded * multiplier
        try:
            if parts[i + 1] == '*':
                if not parts[i + 2].isdigit():
                    raise Exception(f'Unknown command: {" ".join(parts)}')
                expanded += internal_expanded * int(parts[i + 2])
                i += 3
            else:
                expanded += internal_expanded
        except IndexError:
            expanded += internal_expanded
        if not multiplier_set:
            multiplier = 1
        i += 1


class Row:
    def __init__(self, nr, row, stitches):
        self.nr = nr
        self.row = row
        self.stitches = stitches
        self.expanded = []
        self.comment = ''
        stitch_symbols = [x for x, y in stitches]
        tmp = row.split('//', 1)
        if len(tmp) == 2:
            rest, self.comment = tmp
        else:
            rest = tmp[0]
        pattern = f'(([0-9])+|(\()|(\))|\*|#(.*)#|{"|".join(stitch_symbols)})'
        parts = [x[0] for x in re.findall(pattern, rest.strip())]
        expand(parts, stitch_symbols, self.expanded)

    def __len__(self):
        symbols = dict([(x, 1 if y != '++' else 2) for x, y in self.stitches])
        return sum([symbols[x] for x in self.expanded if x in symbols])

    def __str__(self):
        return self.row


class Part:
    def __init__(self, rows, stitches):
        self.name = rows[0].strip(':')
        self.rows = []

        for row_str in rows[1:]:
            row_nr, rest = row_str.strip(' -\n\t').split(':', 1)
            try:
                row_nrs = parse_range(row_nr)
            except ValueError:
                row_nrs = [row_nr]
            for nr in row_nrs:
                self.rows.append(Row(nr, rest, stitches))

    def get_row_count(self):
        return len(self.rows)

    def get_stitch(self, row, i):
        return self.rows[row].expanded[i]

    def __len__(self):
        return sum([len(x) for x in self.rows])

    def __str__(self):
        return self.name


class Pattern:
    def __init__(self, file):
        with open(file, 'r') as fp:
            lines = [l for l in fp.readlines() if l.strip()]
        section = None
        parts = []
        current_indent = 0
        current_part = []
        parts.append(current_part)
        self.stitches = []
        self.part_list = []
        for line in lines:
            if line.lower().startswith('parts:'):
                section = 'parts'
                continue
            if line.lower().startswith('stitches:'):
                section = 'stitches'
                continue
            if line.lower().startswith('part list:'):
                section = 'part_list'
                continue
            if section == 'parts':
                indent = re.search('\\S', line).start()
                if indent < current_indent:
                    current_part = []
                    parts.append(current_part)
                current_indent = indent
                current_part.append(line.strip())
            elif section == 'stitches':
                tmp = line.strip(' -\n\t')
                if "[++]" in tmp:
                    self.stitches.append((tmp.replace("[++]", "").strip(), '++'))
                elif "[--]" in tmp:
                    self.stitches.append((tmp.replace("[--]", "").strip(), '--'))
                else:
                    self.stitches.append((tmp, None))
            elif section == 'part_list':
                nr, part = line.strip(' -\n\t').split(',')
                self.part_list.append((part.strip(), int(nr)))
        self.parts = dict()
        for part in parts:
            p = Part(part, self.stitches)
            self.parts[p.name] = p
            print(p, len(p))

    def __len__(self):
        s = 0
        for part, count in self.part_list:
            s += len(self.parts[part]) * count
        return s


def main():
    print(len(Pattern('patterns/kanin.cro')))
    print(len(Pattern('patterns/bibbi.cro')))


if __name__ == '__main__':
    main()

File: parser.js
const listRIndex = (li, x) => {
    let idx = li.indexOf(x);
    const indices = [];
    while (idx !== -1) {
        indices.push(idx);
        idx = li.indexOf(x, idx + 1);
    }
    return indices[-1];
};

const parseRange = (astr) => {
    const result = new Set();
    astr.split(',').forEach((part) => {
        const x = part.split('-')
        const start = parseInt(x[0]);
        const end = parseInt(x[x.length - 1]) + 1;
        [...Array(end - start).keys()].map(i => i + start).forEach((nr) => result.add(nr));
    });
    return Array.from(result.values()).sort((a, b) => a - b);
}

const expand = (parts, symbols) => {
    const expanded = [];
    let multiplier = 1;
    let i = 0;
    while (i < parts.length) {
        let p = parts[i];
        let internalExpanded = [];
        let multiplierSet = false;
        if (!isNaN(p)) {
            multiplier = parseInt(p, 10)
            multiplierSet = true;
        } else if (symbols.includes(p)) {
            internalExpanded.push(p)
        } else if (p === '(') {
            let index = listRIndex(parts, ')');
            const tmp = expand(parts.slice(i + 1, index), symbols);
            tmp.forEach((k) => internalExpanded.push(k));
            i = index;
        } else if (p.startsWith('#')) {
            internalExpanded.push(`change_color ${p.replace(/^#/, '').trim()}`)
        }
        internalExpanded = Array.from(Array(multiplier)).reduce((acc) => {
            internalExpanded.forEach((j) => {
                acc.push(j);
            })
            return acc;
        }, [])
        if (parts[i + 1] === '*') {
            if (isNaN(parts[i + 2])) {
                throw Error(`Unknown command: ${parts.join(' ')}`)
            }
            internalExpanded = Array.from(Array(parseInt(parts[i + 2], 10))).reduce((acc) => {
                internalExpanded.forEach((i) => {
                    acc.push(i);
                })
                return acc;
            }, [])
            i += 2;
        }
        internalExpanded.forEach((i) => expanded.push(i));
        if (!multiplierSet) {
            multiplier = 1;
        }
        i += 1;
    }
    return expanded;
};

const rowParser = (nr, row, stitches) => {
    const stitchSymbols = stitches.map(([x, y]) => x);
    const data = {
        nr, row, expanded: [], comment: undefined,
    };
    const tmp = row.split('//');
    let rest;
    if (tmp.length >= 2) {
        let comment;
        [rest, ...comment] = tmp;
        data.comment = comment.join('//').trim()
    } else {
        rest = tmp[0]
    }
    const pattern = RegExp(`(([0-9])+|(\\()|(\\))|\\*|#(.*)#|${stitchSymbols.join("|")})`, 'g');
    const parts = Array.from(rest.trim().matchAll(pattern)).map((x) => x[0]);
    data.expanded = expand(parts, stitchSymbols);
    return data;
};

const partParser = (rows, stitches) => {
    const part = {
        name: rows[0].replace(/:/, '').trim(), rows: [],
    };
    rows.forEach((row_str, i) => {
        if (i === 0) {
            return;
        }

        let [row_nr, ...rest] = row_str.trim().replace(/^-/, '').trim().split(':')
        rest = rest.join(':')
        let row_nrs;
        try {
            row_nrs = parseRange(row_nr);
        } catch (e) {
            row_nrs = [row_nr];
        }
        row_nrs.forEach((nr) => {
            if (!rest) {
                return;
            }
            part.rows.push(rowParser(nr, rest, stitches))
        })
    })
    return part;
};

const parse = (data) => {
    const pattern = {
        stitches: [], parts: {}, part_list: []
    };
    let section;
    let currentIndent = 0;
    let currentPart = [];
    const parts = [currentPart];
    data.split('\n').forEach((line) => {
        if (!line.trim()) {
            return;
        }
        if (line.toLowerCase().startsWith('parts:')) {
            section = 'parts'
            return;
        }
        if (line.toLowerCase().startsWith('stitches:')) {
            section = 'stitches'
            return;
        }
        if (line.toLowerCase().startsWith('part list:')) {
            section = 'part_list'
            return;
        }
        if (section === 'parts') {
            const indent = line.search('\\S', line);
            if (indent < currentIndent) {
                currentPart = [];
                parts.push(currentPart);
            }
            currentIndent = indent;
            currentPart.push(line.trim())
        } else if (section === 'stitches') {
            const trimmed_line = line.trim().replace(/^-/, '').trim();
            if (trimmed_line.includes('[++]')) {
                pattern.stitches.push([trimmed_line.replace('[++]', '').trim(), '++'])
            } else if (trimmed_line.includes('[--]')) {
                pattern.stitches.push([trimmed_line.replace('[--]', '').trim(), '++'])
            } else {
                pattern.stitches.push([trimmed_line, undefined]);
            }
        } else if (section === 'part_list') {
            const [nr, part] = line.trim().replace(/^-/, '').trim().split(',');
            pattern.part_list.push([part.trim(), parseInt(nr)]);
        }
    })
    parts.forEach((part) => {
        p = partParser(part, pattern.stitches);
        pattern.parts[p.name] = p;
    });
    return pattern;
};

module.exports = {
    parse,
};
